components/FlexibleHeader/tests/unit/FlexibleHeaderShiftedOffscreenTests.m

```diff
@@ -0,0 +1,148 @@
+
+  // When
+  [fhvc shiftHeaderOffScreenAnimated:NO];
+
+  // Then
+  XCTAssertTrue(fhvc.shiftedOffscreen);
+  XCTAssertEqualWithAccuracy(CGRectGetMinY(fhvc.frame), -fhvc.maximumHeight, 0.001);
+}
+
+- (void)testBecomesTrueWhenShiftedOffscreenAnimatedWithoutTrackingScrollView {
+  // Given
+  MDCFlexibleHeaderView *fhvc = [[MDCFlexibleHeaderView alloc] init];
+
+  // When
+  [fhvc shiftHeaderOffScreenAnimated:YES];
+
+  // Then
+  XCTAssertTrue(fhvc.shiftedOffscreen);
+  XCTAssertEqualWithAccuracy(CGRectGetMinY(fhvc.frame), 0, 0.001);
+}
+
+#pragma mark With a tracking scroll view
+
+- (void)testDefaultIsFalseWithTrackingScrollView {
+  // Given
+  MDCFlexibleHeaderView *fhvc = [[MDCFlexibleHeaderView alloc] init];
+  UIScrollView *scrollView = [[UIScrollView alloc] init];
+  scrollView.frame = CGRectMake(0, 0, 100, 1000);
+  scrollView.contentSize = CGSizeMake(100, 1000);
+  fhvc.trackingScrollView = scrollView;
+
+  // Then
+  XCTAssertFalse(fhvc.shiftedOffscreen);
+  XCTAssertEqualWithAccuracy(CGRectGetMinY(fhvc.frame), 0, 0.001);
+}
+
+- (void)testBecomesTrueWhenShiftedOffscreenAnimatedWithTrackingScrollView {
+  // Given
+  MDCFlexibleHeaderView *fhvc = [[MDCFlexibleHeaderView alloc] init];
+  UIScrollView *scrollView = [[UIScrollView alloc] init];
+  scrollView.frame = CGRectMake(0, 0, 100, 1000);
+  scrollView.contentSize = CGSizeMake(100, 1000);
+  fhvc.trackingScrollView = scrollView;
+
+  // When
+  [fhvc shiftHeaderOffScreenAnimated:YES];
+
+  // Then
+  XCTAssertTrue(fhvc.shiftedOffscreen);
+  XCTAssertEqualWithAccuracy(CGRectGetMinY(fhvc.frame), 0, 0.001);
+}
+
+- (void)testBecomesTrueWhenShiftedOffscreenNotAnimatedWithTrackingScrollView {
+  // Given
+  MDCFlexibleHeaderView *fhvc = [[MDCFlexibleHeaderView alloc] init];
+  UIScrollView *scrollView = [[UIScrollView alloc] init];
+  scrollView.frame = CGRectMake(0, 0, 100, 100);
+  scrollView.contentSize = CGSizeMake(100, 1000);
+  fhvc.trackingScrollView = scrollView;
+
+  // When
+  [fhvc shiftHeaderOffScreenAnimated:NO];
+
+  // Then
+  XCTAssertTrue(fhvc.shiftedOffscreen);
+  XCTAssertEqualWithAccuracy(CGRectGetMinY(fhvc.frame), -fhvc.maximumHeight, 0.001);
+}
+
+#pragma mark - Interactive shifting via trackingScrollViewDidScroll
+
+- (void)testBecomesTrueWhenDraggedOffscreen {
+  // Given
+  MDCFlexibleHeaderView *fhvc = [[MDCFlexibleHeaderView alloc] init];
+  UIScrollView *scrollView = [[UIScrollView alloc] init];
+  scrollView.frame = CGRectMake(0, 0, 100, 100);
+  scrollView.contentSize = CGSizeMake(100, 1000);
+  fhvc.shiftBehavior = MDCFlexibleHeaderShiftBehaviorEnabled;
+  fhvc.trackingScrollView = scrollView;
+  [fhvc trackingScrollViewDidScroll];
+
+  // When
+  scrollView.contentOffset = CGPointMake(0, 400);
+  [fhvc trackingScrollViewDidScroll];
+
+  // Then
+  XCTAssertTrue(fhvc.shiftedOffscreen);
+  XCTAssertEqualWithAccuracy(CGRectGetMinY(fhvc.frame), -fhvc.maximumHeight, 0.001);
+}
+
+- (void)testStaysFalseWhenDraggedPartiallyOffscreen {
```

Comment:
Consider adding a test case (or pair of cases) right at the boundary condition of yes/no for `.shiftedOffscreen` to verify the boundary behaves as expected.  I'm especially curious to understand if the property "waits" to flip until significant change is detected.

For example, imagine the boundary is at 10% with a 5% "hysteresis" margin to avoid rapid flipping:
shifted = NO, offset = 0
offset = 10% of height, shifted = NO
offset = 15% of height, shifted = YES
offset = 10% of height, shifted = YES // hysteresis effect to avoid "flipping" too often at boundary conditions
offset = 5% of height, shifted = NO // Finally flips because exceeded margin beyond trigger point

Reply:
Done. I added `testBecomesFalseWhenScrolledGreaterThanTheHeightOfTheHeaderAndThenBackToTheContentTop` to test going beyond the flipping of the state and then back again.